sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareServiceAsyncClient.java

```diff
@@ -534,4 +538,64 @@ public String generateAccountSas(AccountSasSignatureValues accountSasSignatureVa
         return new AccountSasImplUtil(accountSasSignatureValues)
             .generateSas(SasImplUtils.extractSharedKeyCredential(getHttpPipeline()));
     }
+
+    /**
+     * Restores a previously deleted share.
+     * If the share associated with provided <code>deletedShareName</code>
+     * already exists, this call will result in a 409 (conflict).
```

Comment:
I'm a little confused here, will this undelete the specified share with the name of this share client? I'm guessing this will 409 if we already recreated a share with the delete name because you should rename the share instead?

Reply:
This API is on `ShareServiceAsyncClient` a level up from `ShareAsyncClient` (to have parity with outcome of this discussion https://github.com/Azure/azure-sdk-for-java/pull/10900#discussion_r423420963 for blobs). Therefore `deletedShareName` is the only source of name for restored share. So, if someone created new (or already undeleted) share with same name as `deletedShareName` then restore fails as name is already taken.

The option to rename share while undeleting is not part of this PR due to server side not being ready to release it.